### ISSUE TITLE
missing debug symbols in Visual Studio 2019

### ISSUE THREAD

**Issue by plastovicka** [nidud/asmc#4]
I have this code:
```
symbol1	db "Hello world"
symbol2	db 1,2,3
symbol3	db 4
```
I can see value of symbol3, but cannot see values of symbol1 and symbol2. The debugger shows error message that identifiers are undefined. 
It worked with linker from platform toolset Visual Studio 2017 (v141), but does not work with linker from platform toolset Visual Studio 2019 (v142).
This bug is only in ASMC and JWasm. Other assemblers (MASM, NASM, FASM, POASM) show all symbols correctly.

**Comment by nidud** [comment]
Asmc support symbolic debugging for the CodeView V4 format (same as Jwasm). I've used Visual Studio 2015 for the 64-bit debug samples but haven't really tried anything beyond that. 

There's some more info about this in the dbgcv.h file.
https://github.com/nidud/asmc/blob/master/source/asmc/src/h/dbgcv.h

Think there has been some changes to LINK in VS 2017.

**Comment by nidud** [comment]
blob link: [dbgcv.h]+(../blob/master/source/asmc/src/h/dbgcv.h)
[dbgcv.h](../blob/master/source/asmc/src/h/dbgcv.h)

**Comment by nidud** [comment]
Did some command line testing.
```
    asmc64 -Zi debug_vs.asm
    link -debug -entry:main -subsystem:console debug_vs.obj
Asmc Macro Assembler (x64) Version 2.32.03
 Assembling: debug_vs.asm
Microsoft (R) Incremental Linker Version 14.00.24215.1
debug_vs.obj : fatal error LNK1318: Unexpected PDB error; OK (0) ''
```
This is the same version used for the [test samples](../blob/master/source/test/glu/) and they work just fine from the IDE so that's a bit strange. Using LINK v14.28.29334.0 (VS 2019) seems to work as far as it doesn't produce any errors and the binary work, but I dont know if the symbols are correct.

There's a bug report filed for this issue [fixed in: visual studio 2017 version 15.5](https://developercommunity.visualstudio.com/content/problem/95867/fatal-error-lnk1103-debugging-information-corrupt.html) so this confirm there's a new format used.

**Comment by plastovicka** [comment]
CodeView8 format works in VS 2019. There are already some constants defined in dbgcv.h. 
I can see symbols in the debugger if I use CV8_LF_ARRAY and change all fields in struct cv_typerec_array to uint_32.

**Comment by nidud** [comment]
I used the ./VC/bin/amd64 linker in the test above but the one in ./VC/bin worked fine.

> CodeView8 format works in VS 2019. There are already some constants defined in dbgcv.h.
> I can see symbols in the debugger if I use CV8_LF_ARRAY and change all fields in struct cv_typerec_array to uint_32.
> 

Yes, that's the way. This must however be implemented separately whit the CV4 format as the default but some more testing and documentation is needed.

**Comment by nidud** [comment]
[cv8.txt](https://github.com/yasm/yasm/blob/master/modules/dbgfmts/codeview/cv8.txt) from Yasm.

**Comment by nidud** [comment]
I've started to add some code for CodeView8 and also installed VS 2019 but fail to create a project using assembler.

The implementation will add a command line switch `-cv<4|8>` where cv4 will be the default.

**Comment by nidud** [comment]
Finally manage to produce a binary using VS 2019 so I uploaded the BuildCustomizations files used. The are in the ./source/Asmc directory (vs160.zip).

However, a debug build using ML64 fails using the Win 10.0 SDK so the last working target for symbolic debugging seems to be Win 8.1.

**Comment by nidud** [comment]
Added CodeView 5 and 8 to v2.32.05. Command line switch /Zi5 for CV5 (C11 (vc5.x) 32-bit types) and /Zi8 for CV8 (C13 (vc7.x) zero terminated names). Also added menu items for VS 2019 (vs160.zip).

Currently experimental but the main implementation for testing is at least done.

**Comment by plastovicka** [comment]
The linker shows warning LNK4209: debugging information corrupt; recompile module; linking object as if no debug info.

**Comment by nidud** [comment]
Yes, there's something wrong with the line numbers. It does produce the same debug info as ML64 so it is a bit strange. The only difference I see is a start label added ($$00000) by ML64.

However, the project file for VS 2019 uses Link from .\Hostx64\x86 and that do not throw any errors. The line numbers are still wrong (no source line info or breakpoints) but the symbols loads correctly.

From Windbg CV5 fails to load any symbols but line numbers works. CV8 the opposite: line numbers fails but symbols load correctly.

[codeview.zip](https://github.com/nidud/asmc/files/5643741/codeview.zip)

**Comment by nidud** [comment]
The start label is added. It's a reference to the new line table in CV8 so now it works with all LINK versions for VS 2019. The old line numbers from CV4 and a few other things needed to be removed as well.

There are however a few issues needed to be solved but at least the samples works now.

**Comment by nidud** [comment]
Some improvements

- Added [buffering](../blob/master/source/asmc/src/h/dbgcv.h#L33) to the CV8 sections
- Fixed bug in [pointer size](../blob/master/source/asmc/src/dbgcv.c#L387) (32/64)
- Added [pointer types](../blob/master/source/asmc/src/dbgcv.c#L195) (only void* was used)
- Added more options to [asmc.props](../blob/master/source/test/CV/BuildCustomizations)

The libraries in Asmc can not be used (at least not now) with CV8, so in order to use the include files the INCLUDELIB directive must be stripped from the headers. This is done by the /nolib options which is now added to VS2019 settings. A console sample is added which uses the debug libraries from VS (./test/CV):

```assembly
include stdio.inc

; Add startup module: adjust the path

.pragma comment(linker,"/DEFAULTLIB:\VS19\VC\Tools\MSVC\14.28.29333\lib\x64\libcmtd.lib")

console proto :string_t

    .code

main proc argc:int_t, argv:array_t, environment:ptr string_t
  local string:string_t,
        array[2]:string_t

  ; register arguments is not saved to the stack unless used

    mov eax,argc
    mov rax,argv
    mov rax,environment

  ; load some values to locals

    mov string,&@CStr( "hello world\n" )
    mov array[0],&@CStr( "String 1" )
    mov array[8],&@CStr( "String 2" )

  ; jump to next file

    .return console( string )

main endp

    end
```

**Comment by nidud** [comment]
Struct's didn't work correctly in 64-bit and I added a few more types so it seems to be more stable now. I also added a GUI sample for Windows using the same startup library.

The path to the library from the pragma directive is apparently not needed given this is solved by the IDE so this will work for both Console and Windows applications:
```assembly
.pragma comment(linker,"/DEFAULTLIB:libcmtd.lib")
```

**Comment by nidud** [comment]
The sections in this format may be quite _informative_ so the buffering strategy needed some update. The implementation also failed with more than one module.

This is now tested on a larger project with a substantial amount of information. The [VideoMemory](../blob/master/source/test/DirectX/VideoMemory/) sample have six files which assemble to around 500K each.

**Comment by plastovicka** [comment]
It works on the VideoMemory sample, but it does not work on a Hello world sample:
```
        include windows.inc
        .data
message	db "Hello, world !",0
        .code
main proc
        invoke MessageBox, 0, addr message, addr message, MB_YESNOCANCEL
        ret
main endp
end main
```

**Comment by nidud** [comment]
It works just fine here so what exactly happens?

**Comment by nidud** [comment]
There was a bug in the CV4 implementation (using /Zi). This is fixed in v2.32.11.

**Comment by plastovicka** [comment]
I get warning LNK4209: debugging information corrupt; recompile module; linking object as if no debug info.
I also tried to change linker option /DEBUG to /DEBUG:FULL, but then the linker crashes with fatal error LNK1318: Unexpected PDB error; RPC (23) '(0x000006BA)'.

You can test it on this minimal sample:
```
include windows.inc
.code
a db 'a',0
m: end m
```

If I remove include windows.inc, then it compiles successfully and symbol "a" is visible in the debugger.

**Comment by nidud** [comment]
Another round.

Fixed some of the types and a few other things. It's difficult to say if it's a symbol/type or the size (or count) that creates problems but there seems to be a symbol overload in some cases. Changing 'a' to 'foo' made the symbol visible and extending the hash table made 'a' visible.

In the VideoMemory sample one of the files (VidMemViaWMI) fails but the rest of them works. This is the largest of them so that may be that or some types defined in the headers that is misaligned or rendered incorrectly.

As for the sample this now works:
```assembly
include windows.inc
.code
a db 'a',0
main:
ret
end main

```

**Comment by nidud** [comment]
The symbol-stream was changed for CV8 but alignment was calculated from address and not section size.

This is fixed in v2.32.13 so all the files in the VideoMemory sample works now.

**Comment by plastovicka** [comment]
It's very good. Thank you.
